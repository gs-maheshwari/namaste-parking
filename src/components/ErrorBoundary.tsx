"use client";

import { ErrorInfo, ReactNode } from "react";
import { ErrorBoundary } from "react-error-boundary";

const logError = (error: Error, info: ErrorInfo) => {
  console.error(error, info);
};

interface CustomErrorBoundaryProps {
  children: ReactNode;
}

const CustomErrorBoundary = ({ children }: CustomErrorBoundaryProps) => (
  <ErrorBoundary fallback={<>Something went wrong</>} onError={logError}>
    {children}
  </ErrorBoundary>
);

export default CustomErrorBoundary;
