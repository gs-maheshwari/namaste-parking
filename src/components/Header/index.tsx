"use client";

import type React from "react";
import styles from "./Header.module.css";
import { Button } from "../ui";
import { signOut } from "@/actions";
import { User } from "@/types";

const Header = ({ user }: { user: User }) => {
  return (
    <header className={styles.header}>
      <span className={styles.email}>{user.email}</span>
      <Button variant="outline" onClick={() => signOut()}>
        Sign Out
      </Button>
    </header>
  );
};

export default Header;
