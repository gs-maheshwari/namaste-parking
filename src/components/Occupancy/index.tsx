import styles from "./occupancy.module.css";

interface OccupancyProps {
  title: string;
  occupancy: number;
  capacity: number;
}

const Occupancy = ({ title, occupancy, capacity }: OccupancyProps) => {
  const percentage = (occupancy / capacity) * 100;
  const strokeDasharray = 2 * Math.PI * 45;
  const strokeDashoffset = strokeDasharray * ((100 - percentage) / 100);
  const isHighOccupancy = percentage >= 85;
  
  return (
    <div className={styles.indicator}>
      <h3 className={styles.title}>{title}</h3>
      <div className={styles.circularIndicator}>
        <svg width="120" height="120" viewBox="0 0 120 120">
          <circle
            className={styles.backgroundCircle}
            cx="60"
            cy="60"
            r="45"
            fill="none"
            stroke="#e0e0e0"
            strokeWidth="10"
          />
          <circle
             className={`${styles.progressCircle} ${isHighOccupancy ? styles.highOccupancy : ""}`}
            cx="60"
            cy="60"
            r="45"
            fill="none"
            stroke="#4caf50"
            strokeWidth="10"
            strokeDasharray={strokeDasharray}
            strokeDashoffset={strokeDashoffset}
            transform="rotate(-90 60 60)"
          />
          <text
            x="60"
            y="60"
            textAnchor="middle"
            dy=".3em"
            className={styles.percentageText}
          >
            {percentage.toFixed(0)} %
          </text>
        </svg>
      </div>
      <div className={styles.digital}>
        {occupancy} / {capacity}
      </div>
    </div>
  );
};

export default Occupancy;
