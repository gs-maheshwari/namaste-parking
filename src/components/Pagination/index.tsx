import { useState } from "react";
import styles from "./Pagination.module.css";

interface PaginationProps {
  totalItems: number;
  onPageChange: (offset: number, limit: number) => void;
}

const ITEMS_PER_PAGE = 20;

const Pagination = ({ totalItems, onPageChange }: PaginationProps) => {
  const [currentPage, setCurrentPage] = useState(1);
  const totalPages = Math.ceil(totalItems / ITEMS_PER_PAGE);

  const handlePageChange = (newPage: number) => {
    setCurrentPage(newPage);
    const newOffset = (newPage - 1) * ITEMS_PER_PAGE;
    onPageChange(newOffset, ITEMS_PER_PAGE);
  };

  const handlePrevious = () => {
    if (currentPage > 1) {
      handlePageChange(currentPage - 1);
    }
  };

  const handleNext = () => {
    if (currentPage < totalPages) {
      handlePageChange(currentPage + 1);
    }
  };

  return (
    <div className={styles.pagination}>
      <button
        className={styles.button}
        onClick={handlePrevious}
        disabled={currentPage === 1}
      >
        Previous
      </button>
      <span className={styles.pageInfo}>
        Page {currentPage} of {totalPages}
      </span>
      <button
        className={styles.button}
        onClick={handleNext}
        disabled={currentPage === totalPages}
      >
        Next
      </button>
    </div>
  );
};

export default Pagination;
