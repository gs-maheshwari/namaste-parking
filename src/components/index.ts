export { default as Shell } from './Shell';
export { default as LoginForm } from './LoginForm';
export { default as Overview } from './Overview';
export { default as CustomErrorBoundary } from './ErrorBoundary';
export { default as Sessions } from './Sessions';

export * from './ui'