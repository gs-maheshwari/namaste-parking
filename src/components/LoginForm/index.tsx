"use client";

import { useActionState } from "react";
import { CustomErrorBoundary } from "@/components";
import { signIn } from "@/actions";

import { Input, SubmitButton } from "../ui";
import styles from "./LoginForm.module.css";

const LoginForm = () => {
  const [message, action] = useActionState(signIn, null);

  return (
    <CustomErrorBoundary>
      <section>
        <form action={action} className={styles.form}>
          <h3 className={styles.title}>Sign in</h3>
          <Input
            type="email"
            name="email"
            label="Email"
            placeholder="Email"
            required
          />
          <Input
            type="password"
            name="password"
            label="Password"
            placeholder="Password"
            required
          />
          <SubmitButton>Sign in</SubmitButton>
          {message && <p>{message}</p>}
        </form>
      </section>
    </CustomErrorBoundary>
  );
};

export default LoginForm;
