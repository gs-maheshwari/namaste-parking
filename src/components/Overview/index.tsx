"use server";

import type React from "react";
import styles from "./overview.module.css";
import Occupancy from "../Occupancy";
import { ParkingSpace } from "@/types";

const Overview = ({ parkingSpaces }: { parkingSpaces: ParkingSpace[]}) => {
  const { occupancy: residentsOccupancy = 0, capacity: residentsCapacity = 0 } =
    parkingSpaces.find((space) => space.parkingSpaceId === 1) || {};

  const { occupancy: nrCarsOccupancy = 0, capacity: nrCarsCapacity = 0 } =
    parkingSpaces.find((space) => space.parkingSpaceId === 2) || {};

  const { occupancy: nrMotorsOccupancy = 0, capacity: nrMotorsCapacity = 0 } =
    parkingSpaces.find((space) => space.parkingSpaceId === 3) || {};

  const totalOccupancy = parkingSpaces.reduce(
    (sum, space) => sum + space.occupancy,
    0
  );
  const totalCapacity = parkingSpaces.reduce(
    (sum, space) => sum + space.capacity,
    0
  );

  return (
    <div className={styles.container}>
      <h1 className={styles.title}>Parking Overview</h1>
      <div className={styles.grid}>
        <Occupancy
          title="Residents"
          occupancy={residentsOccupancy}
          capacity={residentsCapacity}
        />
        <Occupancy
          title="Non-Resident Cars"
          occupancy={nrCarsOccupancy}
          capacity={nrCarsCapacity}
        />
        <Occupancy
          title="Non-Resident Motorcycles"
          occupancy={nrMotorsOccupancy}
          capacity={nrMotorsCapacity}
        />
      </div>
      <div className={styles.totalOccupancyWrapper}>
        <div className={styles.totalOccupancy}>
          <h2>Total Occupancy</h2>
          <Occupancy occupancy={totalOccupancy} capacity={totalCapacity} />
        </div>
      </div>
    </div>
  );
};

export default Overview;
