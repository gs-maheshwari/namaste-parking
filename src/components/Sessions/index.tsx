"use client"

import { useMemo, useCallback, useState } from "react"
import useSWR from "swr"
import type { ParkingSession, ParkingSessionsResponse } from "@/types"
import { endParkingSession, getParkingSessions } from "@/services"
import { VehicleType } from "@/lib"

import { Button, Table, TableFilter } from "../ui"
import { DateTime } from "./DateTime"
import Pagination from "../Pagination"

const Sessions = ({ sessionsData }: { sessionsData: ParkingSessionsResponse }) => {
  const [filters, setFilters] = useState<Record<string, string>>({});
  const [endingSessionIds, setEndingSessionIds] = useState<Record<string, boolean>>({})

  const { data, mutate } = useSWR<ParkingSessionsResponse>(
    [filters],
    ([filters]) => getParkingSessions(new URLSearchParams(
      Object.entries(filters)
    ).toString()),
    {
      fallbackData: sessionsData,
      revalidateOnFocus: false,
      revalidateOnReconnect: false,
    },
  )

  const onEndSession = useCallback(
    async (sessionId: string) => {
      setEndingSessionIds((prev) => ({ ...prev, [sessionId]: true }));
      await endParkingSession(sessionId);
      mutate();
      setEndingSessionIds((prev) => ({ ...prev, [sessionId]: false }));
    },
    [mutate],
  )

  const columns = useMemo(
    () => [
      { key: "parkingSessionId", header: "ID" },
      { key: "parkingSpaceId", header: "Space ID" },
      {
        key: "status",
        header: "Status",
        render: (session: ParkingSession) => (session.isSessionEnded ? "Inactive" : "Active"),
      },
      {
        key: "sessionLengthInHoursMinutes",
        header: "Length",
        render: (session: ParkingSession) => `${convertToHoursMinutes(session.sessionLengthInHoursMinutes || 0)}`,
      },
      {
        key: "sessionStartedAt",
        header: "Start Time",
        render: (session: ParkingSession) => <DateTime isoDate={session.sessionStartedAt} />,
      },
      {
        key: "sessionEndedAt",
        header: "End Time",
        render: (session: ParkingSession) =>
          session.sessionEndedAt ? <DateTime isoDate={session.sessionEndedAt} /> : "N/A",
      },
      { key: "vehicleLicensePlate", header: "License Plate" },
      { key: "vehicleType", header: "Vehicle Type" },
      {
        key: "revenue",
        header: "Revenue",
        render: (session: ParkingSession) =>
          session.sessionEndedAt && session.parkingSpaceId !== 1
            ? `â‚¬ ${getRevenue(session.sessionLengthInHoursMinutes || 0, session.vehicleType)}`
            : "N/A",
      },
      {
        key: "actions",
        header: "Actions",
        render: (session: ParkingSession) =>
          !session.isSessionEnded && (
            <Button onClick={() => onEndSession(session.parkingSessionId)} disabled={endingSessionIds[session.parkingSessionId]} variant="primary">
              {endingSessionIds[session.parkingSessionId] ? "Ending..." : "End"}
            </Button>
          ),
      },
    ],
    [onEndSession, endingSessionIds],
  )

  const filterConfig = [
    {
      key: "vehicleType",
      type: "select" as const,
      label: "Vehicle Type",
      options: [
        { value: VehicleType.CAR, label: "CAR" },
        { value: VehicleType.MOTOR, label: "MOTOR" },
      ],
    },
    { key: "sessionStartedAtFrom", type: "date" as const, label: "Start Date" },
    { key: "sessionEndedAtTo", type: "date" as const, label: "End Date" },
    {
      key: "isSessionEnded",
      type: "select" as const,
      label: "Status",
      options: [
        { value: "false", label: "Active" },
        { value: "true", label: "Inactive" },
      ],
    },
  ]

  const onFilterChange = useCallback((newFilters: Record<string, string>) => {
    setFilters(newFilters)
  }, [])

  return (
    <>
      <TableFilter filters={filters} filterConfig={filterConfig} onFilterChange={onFilterChange} />
      {data?.parkingSessions && <Table data={data.parkingSessions} columns={columns} />}
      {data?.parkingSessionsTotalCount || 0 > 0 ? (
        <Pagination
          totalItems={data?.parkingSessionsTotalCount || 0}
          onPageChange={(newOffset, limit) =>
            setFilters((prev) => ({
              ...prev,
              offset: newOffset.toString(),
              limit: limit.toString(),
            }))
          }
        />
      ) : null}
    </>
  )
}

const convertToHoursMinutes = (totalMinutes: number) => {
  const hours = Math.floor(totalMinutes / 60)
  const minutes = totalMinutes % 60
  return `${hours}h ${minutes}m`
}

const getRevenue = (totalMinutes: number, vehicleType: VehicleType) => {
  const hours = Math.ceil(totalMinutes / 60)
  const revenue = vehicleType === VehicleType.CAR ? hours * 5 : hours * 3
  return revenue
}

export default Sessions

