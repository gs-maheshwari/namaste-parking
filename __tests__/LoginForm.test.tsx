import { render } from "@testing-library/react";
import { LoginForm } from "../src/components";

describe("LoginForm", () => {
  it("renders LoginForm unchanged", () => {
    const { container } = render(<LoginForm />);
    expect(container).toMatchSnapshot();
  });
});
