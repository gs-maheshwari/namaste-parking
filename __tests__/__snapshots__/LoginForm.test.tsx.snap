// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LoginForm renders LoginForm unchanged 1`] = `
<div>
  <section>
    <form
      action="javascript:throw new Error('A React form was unexpectedly submitted. If you called form.submit() manually, consider using form.requestSubmit() instead. If you\\'re trying to use event.stopPropagation() in a submit event handler, consider also calling event.preventDefault().')"
      class="form"
    >
      <h3
        class="title"
      >
        Sign in
      </h3>
      <div
        class="inputWrapper"
      >
        <label
          class="label"
        >
          Email
        </label>
        <input
          class="input"
          name="email"
          placeholder="Email"
          required=""
          type="email"
        />
      </div>
      <div
        class="inputWrapper"
      >
        <label
          class="label"
        >
          Password
        </label>
        <input
          class="input"
          name="password"
          placeholder="Password"
          required=""
          type="password"
        />
      </div>
      <button
        class="button primary"
        type="submit"
      >
        Sign in
      </button>
    </form>
  </section>
</div>
`;
